### PR TITLE
Creature spellcast refactor

### PR DESCRIPTION
**Pull request by dydzio0614**
- Somewhat improved cursor handling for battle hexes for client
- Spellcast is now default action for stacks able to perform it (bonus NO_SPELLCAST_BY_DEFAULT negates that)
- Creatures with damaging spells are now able to use them if controlled by BattleAI

Things I am not sure about:
- new method CBattleInfoCallback::getClientActionsForStack - is it proper place for this code?
- what to do with StupidAI spellcasting?

### CODE REVIEW

File: AI/BattleAI/BattleAI.cpp [hunk 1]
```diff
@@ -521,6 +555,64 @@ void CBattleAI::attemptCastingSpell()
 	}
 }
 
+void CBattleAI::evaluateCreatureSpellcast(const CStack * stack, PossibleSpellcast & ps)
```
[the remark this answers is not in the document]
Comment (dydzio0614) [reply AI/BattleAI/BattleAI.cpp]: valueOfStack and newValueOfStack are actually used, so how should I cleanup them?

[the remark this answers is not in the document]
Comment (alexvins) [reply AI/BattleAI/BattleAI.cpp]: Am I blind? I see only zero initialization. In my code there is a complex calculation fro them, but you only evaluate damage.

[the remark this answers is not in the document]
Comment (dydzio0614) [reply AI/BattleAI/BattleAI.cpp]: Check code lines 596-598

[the remark this answers is not in the document]
Comment (dydzio0614) [reply AI/BattleAI/BattleAI.cpp]: Wait, seems you're right...

**Comment by dydzio0614** [comment]
VS 2015 build has problem with using boost::optional the "currently intended way" according to new boost version wiki pages. I suggest it may be good time for breaking VS 2015 compatibility. It will have to be  broken anyway if we plan to migrate to newer C++ standard soon.

**Comment by dydzio0614** [comment]
What is left to get this branch merged?

**Pull request merged**